### PR TITLE
feat(integrations): Save dynamic field choices to the rule data

### PR DESCRIPTION
**Pull request by mgaeta**
The Ticket Rule form gets its fields from Jira's "createmeta" API. Every non-user field has its choices included but `assignee` and `reporter` need to get users in a separate API call.
This PR captures those user choices and saves them to the form's choices.

### CODE REVIEW

File: src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx [hunk 1]
```diff
@@ -107,10 +112,20 @@ class TicketRuleForm extends React.Component<Props, State> {
         const options = choices.map(([value, label]) => ({value, label}));
         return resolve({options});
       }
+
       return this.debouncedOptionLoad(field, input, (err, result) => {
         if (err) {
           reject(err);
         } else {
+          const dynamicFieldChoices = result.options.map(obj => [obj.value, obj.label]);
+          this.setState(state => {
```
Comment (mgaeta) [inline src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx:121]: @NisanthanNanthakumar is there a better way to do this?

File: src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ruleNode.tsx [hunk 2]
```diff
@@ -277,11 +277,26 @@ class RuleNode extends React.Component<Props> {
...
+    data: {[key: string]: string},
+    dynamicFieldChoices: {[key: string]: string[]}
+  ): void => {
     // iterating through these upon save instead of when each
     // element is changed to match the spec
     for (const [name, value] of Object.entries(data)) {
       this.props.onPropertyChange(this.props.index, name, value);
+
+      // We only know the choices after the form loads.
+      if (name in ['assignee', 'reporter'] && dynamicFieldChoices[name]) {
```
Comment (scttcper) [inline src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ruleNode.tsx:290]:
```suggestion
      if (['assignee', 'reporter'].includes(name) && dynamicFieldChoices[name]) {
```

File: src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx (hunk 1 above)
Comment (NisanthanNanthakumar) [reply src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx:121]: @mgaeta the spread operator on objects returns a shallow copy, which is not ideal for  multidimensional objects. We can use the `cloneDeep` and `set` functions from `lodash` to make a deep clone, update the cloned object and then setState with that object.

```suggestion
import cloneDeep from 'lodash/cloneDeep';
import set from 'lodash/set';

this.setState(prevState => {
  const newState = cloneDeep(prevState);
  set(newState, `dynamicFieldChoices[${field.name}]`, value);
  return newState;
});
```

File: src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx [hunk 3]
```diff
@@ -19,18 +19,23 @@ import {FormField} from 'app/views/settings/projectAlerts/issueEditor/ruleNode';
...
+    data: {[key: string]: string},
+    dynamicFieldChoices: {[key: string]: string[]}
+  ) => void;
   onPropertyChange: (rowIndex: number, name: string, value: string) => void;
   index: number;
 };
 
 type State = {
   showModal: boolean;
+  dynamicFieldChoices: any;
```
Comment (NisanthanNanthakumar) [inline src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx:32]: @mgaeta we can probably be stricter with the types. We can create a new type:
``` type dynamicFieldChoices = {[key: string]: string[]}``` and use this type instead.

File: src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx (hunk 1 above)
Comment (mgaeta) [reply src/sentry/static/sentry/app/views/settings/projectAlerts/issueEditor/ticketRuleForm.tsx:121]: 👍🏽

**Review by NisanthanNanthakumar — APPROVED** [review]

**Pull request merged**